### ISSUE TITLE
v2.17.0 compile error with LuaJIT 2.1.0-beta2 on msys2 mingw64 g++ 6.3.0

### ISSUE THREAD

**Issue by ketigid** [ThePhD/sol2#392]
Hi,

I got a long list of compile errors:

```
sol.hpp: In constructor 'sol::absolute_index::absolute_index(lua_State*, int)':
sol.hpp:3182:68: error: 'lua_absindex' was not declared in this scope
   absolute_index(lua_State* L, int idx) : index(lua_absindex(L, idx)) {}
                                                                    ^
sol.hpp: At global scope:
sol.hpp:3345:8: error: 'LUA_OK' was not declared in this scope
   ok = LUA_OK,
        ^~~~~~
sol.hpp:3350:8: error: 'LUA_ERRGCMM' was not declared in this scope
   gc = LUA_ERRGCMM,
        ^~~~~~~~~~~
sol.hpp:3356:8: error: 'LUA_OK' was not declared in this scope
   ok = LUA_OK,
        ^~~~~~
sol.hpp:3360:8: error: 'LUA_ERRGCMM' was not declared in this scope
   gc = LUA_ERRGCMM,
        ^~~~~~~~~~~
sol.hpp:3366:8: error: 'LUA_OK' was not declared in this scope
   ok = LUA_OK,
        ^~~~~~
sol.hpp:3369:8: error: 'LUA_ERRGCMM' was not declared in this scope
   gc = LUA_ERRGCMM,
        ^~~~~~~~~~~
sol.hpp: In function 'const string& sol::to_string(sol::call_status)':
sol.hpp:3417:3: error: duplicate case value
   case call_status::file:
   ^~~~
sol.hpp:3413:3: error: previously used here
   case call_status::gc:
   ^~~~
sol.hpp: In constructor 'sol::stack_reference::stack_reference(lua_State*, int)':
sol.hpp:3955:80: error: 'lua_absindex' was not declared in this scope
   stack_reference(lua_State* L, int i) noexcept : L(L), index(lua_absindex(L, i)) {}
```
There's more but I truncated it.
It compiled fine on v2.16

**Comment by ThePhD** [comment]
It looks like it doesn't have certain constants define, and certain functions. This is bizarre... I'll look into why it's happening later.

**Comment by ThePhD** [comment]
@ketigid No repro. Everything I have compiles and runs using LuaJIT 2.1.0-beta2 from LuaJIT's website after I build the DLLs and all. You need to examine what version of Sol you're using, and probably show me the full build log dumped in a gist or something so I can get a better understanding of what's going wrong.

If you configure LuaJIT's headers in any specific what, it'd be good to know that too.

**Comment by ketigid** [comment]
I've just reproduced it on a different Win7 machine.
Here are the steps I took.
- Install msys2
- Install mingw-w64-x86_64-gcc, it turns out to be version 6.3.0
- Install GNU make 4.2.1
- Open up mingw64, the 64-bit environment
- Build LuaJIT-2.1.0-beta2 by running `make` (no changes made)
- Downloaded `sol.hpp` (release v2.17.0) to ~/
- Create a sample `hello.cpp`
```
#include <sol.hpp>

int main() {
    sol::state lua;
    lua.open_libraries();
    lua.script("print('bark!')");
}
```
- Run `g++ hello.cpp -o hello.exe -I. -I./LuaJIT-2.1.0-beta2/src -L./LuaJIT-2.1.0-beta2/src -llua51`
- Error! See full build log below.
- Next, I downloaded `sol.hpp` (release v2.16.0) to ~/
- Run the same build command `g++ hello.cpp -o hello.exe -I. -I./LuaJIT-2.1.0-beta2/src -L./LuaJIT-2.1.0-beta2/src -llua51`
- Success! `bark!`

Here's the full build log.
```
In file included from hello.cpp:1:0:
./sol.hpp: In constructor 'sol::absolute_index::absolute_index(lua_State*, int)':
./sol.hpp:3182:68: error: 'lua_absindex' was not declared in this scope
   absolute_index(lua_State* L, int idx) : index(lua_absindex(L, idx)) {}
                                                                    ^
./sol.hpp: At global scope:
./sol.hpp:3345:8: error: 'LUA_OK' was not declared in this scope
   ok = LUA_OK,
        ^~~~~~
./sol.hpp:3350:8: error: 'LUA_ERRGCMM' was not declared in this scope
   gc = LUA_ERRGCMM,
        ^~~~~~~~~~~
./sol.hpp:3356:8: error: 'LUA_OK' was not declared in this scope
   ok = LUA_OK,
        ^~~~~~
./sol.hpp:3360:8: error: 'LUA_ERRGCMM' was not declared in this scope
   gc = LUA_ERRGCMM,
        ^~~~~~~~~~~
./sol.hpp:3366:8: error: 'LUA_OK' was not declared in this scope
   ok = LUA_OK,
        ^~~~~~
./sol.hpp:3369:8: error: 'LUA_ERRGCMM' was not declared in this scope
   gc = LUA_ERRGCMM,
        ^~~~~~~~~~~
./sol.hpp: In function 'const string& sol::to_string(sol::call_status)':
./sol.hpp:3417:3: error: duplicate case value
   case call_status::file:
   ^~~~
./sol.hpp:3413:3: error: previously used here
   case call_status::gc:
   ^~~~
./sol.hpp: In constructor 'sol::stack_reference::stack_reference(lua_State*, int)':
./sol.hpp:3955:80: error: 'lua_absindex' was not declared in this scope
   stack_reference(lua_State* L, int i) noexcept : L(L), index(lua_absindex(L, i)) {}
                                                                                ^
./sol.hpp: In function 'bool sol::operator==(const sol::stack_reference&, const sol::stack_reference&)':
./sol.hpp:4004:71: error: 'LUA_OPEQ' was not declared in this scope
   return lua_compare(l.lua_state(), l.stack_index(), r.stack_index(), LUA_OPEQ) == 0;
                                                                       ^~~~~~~~
./sol.hpp:4004:79: error: 'lua_compare' was not declared in this scope
   return lua_compare(l.lua_state(), l.stack_index(), r.stack_index(), LUA_OPEQ) == 0;
                                                                               ^
./sol.hpp: In constructor 'sol::reference::reference(lua_State*, sol::detail::global_tag)':
./sol.hpp:4082:35: error: 'lua_pushglobaltable' was not declared in this scope
    lua_pushglobaltable(lua_state());
                                   ^
./sol.hpp: In function 'bool sol::operator==(const sol::reference&, const sol::reference&)':
./sol.hpp:4188:45: error: 'LUA_OPEQ' was not declared in this scope
   return lua_compare(l.lua_state(), -1, -2, LUA_OPEQ) == 1;
                                             ^~~~~~~~
./sol.hpp:4188:53: error: 'lua_compare' was not declared in this scope
   return lua_compare(l.lua_state(), -1, -2, LUA_OPEQ) == 1;
                                                     ^
In file included from hello.cpp:1:0:
./sol.hpp: In static member function 'static bool sol::stack::checker<T, (sol::type)3, typename std::enable_if<std::is_integral<_Tp>::value, void>::type>::check(lua_State*, int, Handler&&, sol::stack::record&)':
./sol.hpp:5036:42: error: there are no arguments to 'lua_isinteger' that depend on a template parameter, so a declaration of 'lua_isinteger' must be available [-fpermissive]
     bool success = lua_isinteger(L, index) == 1;
                                          ^
./sol.hpp:5036:42: note: (if you use '-fpermissive', G++ will accept your code, but allowing the use of an undeclared name is deprecated)
In file included from hello.cpp:1:0:
./sol.hpp: In static member function 'static T sol::stack::getter<T, typename std::enable_if<std::is_enum<_Tp>::value, void>::type>::get(lua_State*, int, sol::stack::record&)':
./sol.hpp:5404:59: error: there are no arguments to 'lua_tointegerx' that depend on a template parameter, so a declaration of 'lua_tointegerx' must be available [-fpermissive]
     return static_cast<T>(lua_tointegerx(L, index, nullptr));
                                                           ^
./sol.hpp: In static member function 'static T sol::stack::getter<sol::as_table_t<T>, typename std::enable_if<(! sol::meta::has_key_value_pair<typename std::remove_cv<typename std::remove_reference<_From>::type>::type>::value), void>::type>::get(sol::types<U>, lua_State*, int, sol::stack::record&)':
./sol.hpp:5419:41: error: there are no arguments to 'lua_absindex' that depend on a template parameter, so a declaration of 'lua_absindex' must be available [-fpermissive]
     int index = lua_absindex(L, relindex);
                                         ^
./sol.hpp: In static member function 'static T sol::stack::getter<sol::as_table_t<T>, typename std::enable_if<sol::meta::has_key_value_pair<typename std::remove_cv<typename std::remove_reference<_From>::type>::type>::value, void>::type>::get(sol::types<K, V>, lua_State*, int, sol::stack::record&)':
./sol.hpp:5480:41: error: there are no arguments to 'lua_absindex' that depend on a template parameter, so a declaration of 'lua_absindex' must be available [-fpermissive]
     int index = lua_absindex(L, relindex);
                                         ^
In file included from hello.cpp:1:0:
./sol.hpp: In static member function 'static sol::optional<T> sol::stack::check_getter<T, typename std::enable_if<(std::is_integral<_Tp>::value && (sol::lua_type_of<T>::value == number)), void>::type>::get(lua_State*, int, Handler&&, sol::stack::record&)':
./sol.hpp:5969:56: error: there are no arguments to 'lua_tointegerx' that depend on a template parameter, so a declaration of 'lua_tointegerx' must be available [-fpermissive]
     lua_Integer value = lua_tointegerx(L, index, &isnum);
                                                        ^
./sol.hpp: In static member function 'static sol::optional<T> sol::stack::check_getter<T, typename std::enable_if<(std::is_enum<_Tp>::value && (! sol::meta::any_same<T, sol::meta_function, sol::type>::value)), void>::type>::get(lua_State*, int, Handler&&, sol::stack::record&)':
./sol.hpp:5986:56: error: there are no arguments to 'lua_tointegerx' that depend on a template parameter, so a declaration of 'lua_tointegerx' must be available [-fpermissive]
     lua_Integer value = lua_tointegerx(L, index, &isnum);
                                                        ^
./sol.hpp: In static member function 'static sol::optional<T> sol::stack::check_getter<T, typename std::enable_if<std::is_floating_point<_Tp>::value, void>::type>::get(lua_State*, int, Handler&&, sol::stack::record&)':
./sol.hpp:6003:54: error: there are no arguments to 'lua_tonumberx' that depend on a template parameter, so a declaration of 'lua_tonumberx' must be available [-fpermissive]
     lua_Number value = lua_tonumberx(L, index, &isnum);
                                                      ^
./sol.hpp: In member function 'void sol::stack::field_getter<std::tuple<_Elements ...>, b, raw, C>::get(lua_State*, Keys&&)':
./sol.hpp:6918:103: error: there are no arguments to 'lua_absindex' that depend on a template parameter, so a declaration of 'lua_absindex' must be available [-fpermissive]
     apply(std::make_index_sequence<sizeof...(Args)>(), L, std::forward<Keys>(keys), lua_absindex(L, -1));
                                                                                                       ^
In file included from hello.cpp:1:0:
./sol.hpp: In function 'sol::call_syntax sol::stack::get_call_syntax(lua_State*, const string&, int)':
./sol.hpp:7284:34: error: 'LUA_OPEQ' was not declared in this scope
    if (lua_compare(L, -1, index, LUA_OPEQ) != 1) {
                                  ^~~~~~~~
./sol.hpp:7284:42: error: 'lua_compare' was not declared in this scope
    if (lua_compare(L, -1, index, LUA_OPEQ) != 1) {
                                          ^
In file included from hello.cpp:1:0:
./sol.hpp: In constructor 'sol::variadic_args::variadic_args(lua_State*, int)':
./sol.hpp:10014:113: error: 'lua_absindex' was not declared in this scope
   variadic_args(lua_State* luastate, int stackindex = -1) : L(luastate), index(lua_absindex(luastate, stackindex)), stacktop(lua_gettop(luastate)) {}
                                                                                                                 ^
In file included from hello.cpp:1:0:
./sol.hpp: In function 'int sol::usertype_detail::metatable_newindex(lua_State*)':
./sol.hpp:10523:66: error: there are no arguments to 'lua_tointegerx' that depend on a template parameter, so a declaration of 'lua_tointegerx' must be available [-fpermissive]
    lua_Integer magic = lua_tointegerx(L, upvalue_index(4), &isnum);
                                                                  ^
./sol.hpp: In static member function 'static int sol::stack::pusher<sol::usertype_metatable<T, std::integer_sequence<long long unsigned int, I ...>, Tn ...> >::push(lua_State*, sol::stack::pusher<sol::usertype_metatable<T, std::integer_sequence<long long unsigned int, I ...>, Tn ...> >::umt_t&&)':
./sol.hpp:10983:34: error: there are no arguments to 'luaL_setfuncs' that depend on a template parameter, so a declaration of 'luaL_setfuncs' must be available [-fpermissive]
      luaL_setfuncs(L, metaregs, 1);
                                  ^
In file included from hello.cpp:1:0:
./sol.hpp: In static member function 'static int sol::stack::pusher<sol::simple_usertype_metatable<T> >::push(lua_State*, sol::stack::pusher<sol::simple_usertype_metatable<T> >::umt_t&& ':
./sol.hpp:11472:32: error: there are no arguments to 'luaL_setfuncs' that depend on a template parameter, so a declaration of 'luaL_setfuncs' must be available [-fpermissive]
      luaL_setfuncs(L, opregs, 0);
                                ^
./sol.hpp: In member function 'void sol::state_view::open_libraries(Args&& ...)':
./sol.hpp:13001:46: error: there are no arguments to 'luaL_requiref' that depend on a template parameter, so a declaration of 'luaL_requiref' must be available [-fpermissive]
      luaL_requiref(L, "base", luaopen_base, 1);
                                              ^
./sol.hpp:13005:52: error: there are no arguments to 'luaL_requiref' that depend on a template parameter, so a declaration of 'luaL_requiref' must be available [-fpermissive]
      luaL_requiref(L, "package", luaopen_package, 1);
                                                    ^
./sol.hpp:13017:50: error: there are no arguments to 'luaL_requiref' that depend on a template parameter, so a declaration of 'luaL_requiref' must be available [-fpermissive]
      luaL_requiref(L, "string", luaopen_string, 1);
                                                  ^
./sol.hpp:13021:48: error: there are no arguments to 'luaL_requiref' that depend on a template parameter, so a declaration of 'luaL_requiref' must be available [-fpermissive]
      luaL_requiref(L, "table", luaopen_table, 1);
                                                ^
./sol.hpp:13025:46: error: there are no arguments to 'luaL_requiref' that depend on a template parameter, so a declaration of 'luaL_requiref' must be available [-fpermissive]
      luaL_requiref(L, "math", luaopen_math, 1);
                                              ^
./sol.hpp:13030:46: error: there are no arguments to 'luaL_requiref' that depend on a template parameter, so a declaration of 'luaL_requiref' must be available [-fpermissive]
      luaL_requiref(L, "bit32", luaopen_bit, 1);
                                              ^
./sol.hpp:13039:42: error: there are no arguments to 'luaL_requiref' that depend on a template parameter, so a declaration of 'luaL_requiref' must be available [-fpermissive]
      luaL_requiref(L, "io", luaopen_io, 1);
                                          ^
./sol.hpp:13043:42: error: there are no arguments to 'luaL_requiref' that depend on a template parameter, so a declaration of 'luaL_requiref' must be available [-fpermissive]
      luaL_requiref(L, "os", luaopen_os, 1);
                                          ^
./sol.hpp:13047:48: error: there are no arguments to 'luaL_requiref' that depend on a template parameter, so a declaration of 'luaL_requiref' must be available [-fpermissive]
      luaL_requiref(L, "debug", luaopen_debug, 1);
                                                ^
./sol.hpp:13058:44: error: there are no arguments to 'luaL_requiref' that depend on a template parameter, so a declaration of 'luaL_requiref' must be available [-fpermissive]
      luaL_requiref(L, "ffi", luaopen_ffi, 1);
                                            ^
./sol.hpp:13064:44: error: there are no arguments to 'luaL_requiref' that depend on a template parameter, so a declaration of 'luaL_requiref' must be available [-fpermissive]
      luaL_requiref(L, "jit", luaopen_jit, 1);
                                            ^
./sol.hpp: In member function 'sol::object sol::state_view::require(const string&, lua_CFunction, bool)':
./sol.hpp:13076:70: error: 'luaL_requiref' was not declared in this scope
    luaL_requiref(L, key.c_str(), open_function, create_global ? 1 : 0);
                                                                      ^
./sol.hpp: In member function 'sol::load_result sol::state_view::load(const string&)':
./sol.hpp:13190:44: error: 'lua_absindex' was not declared in this scope
    return load_result(L, lua_absindex(L, -1), 1, 1, x);
                                            ^
./sol.hpp: In member function 'sol::load_result sol::state_view::load_file(const string&)':
./sol.hpp:13195:44: error: 'lua_absindex' was not declared in this scope
    return load_result(L, lua_absindex(L, -1), 1, 1, x);
                                            ^
./sol.hpp: In member function 'sol::load_result sol::state_view::load_buffer(const char*, size_t, const char*, const char*)':
./sol.hpp:13200:44: error: 'lua_absindex' was not declared in this scope
    return load_result(L, lua_absindex(L, -1), 1, 1, x);
                                            ^
./sol.hpp: In member function 'sol::protected_function_result sol::coroutine::invoke(sol::types<>, std::index_sequence<>, std::ptrdiff_t)':
./sol.hpp:13664:79: error: 'lua_absindex' was not declared in this scope
     return protected_function_result(lua_state(), lua_absindex(lua_state(), -1), 1, returncount, status());
                                                                               ^
./sol.hpp: In instantiation of 'void sol::state_view::open_libraries(Args&& ...) [with Args = {}]':
hello.cpp:5:24:   required from here
./sol.hpp:13001:19: error: 'luaL_requiref' was not declared in this scope
      luaL_requiref(L, "base", luaopen_base, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13005:19: error: 'luaL_requiref' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
      luaL_requiref(L, "package", luaopen_package, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13001:19: note: 'luaL_requiref' declared here, later in the translation unit
      luaL_requiref(L, "base", luaopen_base, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13017:19: error: 'luaL_requiref' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
      luaL_requiref(L, "string", luaopen_string, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13001:19: note: 'luaL_requiref' declared here, later in the translation unit
      luaL_requiref(L, "base", luaopen_base, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13021:19: error: 'luaL_requiref' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
      luaL_requiref(L, "table", luaopen_table, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13001:19: note: 'luaL_requiref' declared here, later in the translation unit
      luaL_requiref(L, "base", luaopen_base, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13025:19: error: 'luaL_requiref' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
      luaL_requiref(L, "math", luaopen_math, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13001:19: note: 'luaL_requiref' declared here, later in the translation unit
      luaL_requiref(L, "base", luaopen_base, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13030:19: error: 'luaL_requiref' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
      luaL_requiref(L, "bit32", luaopen_bit, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13001:19: note: 'luaL_requiref' declared here, later in the translation unit
      luaL_requiref(L, "base", luaopen_base, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13039:19: error: 'luaL_requiref' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
      luaL_requiref(L, "io", luaopen_io, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13001:19: note: 'luaL_requiref' declared here, later in the translation unit
      luaL_requiref(L, "base", luaopen_base, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13043:19: error: 'luaL_requiref' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
      luaL_requiref(L, "os", luaopen_os, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13001:19: note: 'luaL_requiref' declared here, later in the translation unit
      luaL_requiref(L, "base", luaopen_base, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13047:19: error: 'luaL_requiref' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
      luaL_requiref(L, "debug", luaopen_debug, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13001:19: note: 'luaL_requiref' declared here, later in the translation unit
      luaL_requiref(L, "base", luaopen_base, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13058:19: error: 'luaL_requiref' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
      luaL_requiref(L, "ffi", luaopen_ffi, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13001:19: note: 'luaL_requiref' declared here, later in the translation unit
      luaL_requiref(L, "base", luaopen_base, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13064:19: error: 'luaL_requiref' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
      luaL_requiref(L, "jit", luaopen_jit, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
./sol.hpp:13001:19: note: 'luaL_requiref' declared here, later in the translation unit
      luaL_requiref(L, "base", luaopen_base, 1);
      ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from hello.cpp:1:0:
./sol.hpp: In instantiation of 'sol::call_status sol::basic_protected_function<T>::luacall(std::ptrdiff_t, std::ptrdiff_t, sol::detail::handler&) const [with base_t = sol::reference; std::ptrdiff_t = long long int]':
./sol.hpp:9699:19:   required from 'sol::protected_function_result sol::basic_protected_function<T>::invoke(sol::types<>, std::index_sequence<>, std::ptrdiff_t, sol::detail::handler&) const [with base_t = sol::reference; std::index_sequence<> = std::integer_sequence<long long unsigned int>; std::ptrdiff_t = long long int]'
./sol.hpp:9778:17:   required from 'decltype(auto) sol::basic_protected_function<T>::call(Args&& ...) const [with Ret = {}; Args = {}; base_t = sol::reference]'
./sol.hpp:9765:17:   required from 'sol::protected_function_result sol::basic_protected_function<T>::operator()(Args&& ...) const [with Args = {}; base_t = sol::reference]'
./sol.hpp:13116:14:   required from here
./sol.hpp:9658:46: error: 'lua_pcallk' was not declared in this scope
    return static_cast<call_status>(lua_pcallk(base_t::lua_state(), static_cast<int>(argcount), static_cast<int>(resultcount), h.stackindex, 0, nullptr));
                                    ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by ThePhD** [comment]
It's not finding the right things (and putting out the wrong version) because you're not pointing it to a proper `lua.hpp`. sol2 includes Lua using `lua.hpp` (or all the relevant includes if it detects you're building with C++ mangling, but you have to specify that manually through a preprocessor directive). LuaJIT-2.1.0-beta2 doesn't have lua.hpp, and your error log tells me that it's getting included from somewhere else, probably with all the wrong versions being used.

When I run that command, my code fails with a `fatal error: lua.hpp: No such file or directory`. Please make sure your build path or system paths aren't contaminated with bad Lua's, or add a proper lua.hpp somewhere in the LuaJIT directory. Mine looks like this:

````
// lua.hpp
// Lua header files for C++
// <<extern "C">> not supplied automatically because Lua also compiles as C++

extern "C" {
#include "lua.h"
#include "lualib.h"
#include "lauxlib.h"
}
````

**Comment by ThePhD** [comment]
Well, I finally managed to reproduce it after hammering some things down. The problem is, I can't fix it. =/

At least, at the moment. I've analyzed the preprocessor defines of MinGW-x86_64 for g++ 6.3.0 for the simple file, and its showing not only LUA_VERSION_NUM defined at 501, but LUAJIT_VERSION defined as well.

Then, I examine the hello.cpp, but with all of the code dumped out in a single file (standard library code, sol code, and all). As it turns out, the preprocessed source is not a fairly large chunk of the Sol compatibility code, despite the defines checking out. I'm still digging through the 135,000+ line file to figure out WTF MinGW g++ 6.3.0 is on to make this kind of goof up, but at the very least it's there.
